### PR TITLE
feat: create gatehub user

### PR DESCRIPTION
**Pull request by dragosp1011**
### Context

- fixes #

### CODE REVIEW

File: packages/wallet/backend/src/gatehub/client.ts [hunk 1]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
-      DEFAULT_APP_SCOPE
+      DEFAULT_APP_SCOPE,
+      managerUserId
     )
 
     return `${this.rampUrl}/?paymentType=${PAYMENT_TYPE.deposit}&bearer=${token}`
   }
 
-  async getOnboardingUrl(): Promise<string> {
+  async getOnboardingUrl(managerUserId: string): Promise<string> {
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:88]: There are multiple places where the variable is named `managerUserId`. I'll try to point out all of them:
```suggestion
  async getOnboardingUrl(managedUserId: string): Promise<string> {
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 2]
```diff
@@ -26,7 +30,10 @@ export class GateHubClient {
   private clientIds = SANDBOX_CLIENT_IDS
   private mainUrl = 'sandbox.gatehub.net'
 
-  private iframeMappings: Record<IFRAME_TYPE, () => Promise<string>> = {
+  private iframeMappings: Record<
+    IFRAME_TYPE,
+    (managerUserId: string) => Promise<string>
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:35]:
```suggestion
    (managedUserId: string) => Promise<string>
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 3]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
     return `https://onboarding.${this.mainUrl}`
   }
 
-  async getWithdrawalUrl(): Promise<string> {
+  async getWithdrawalUrl(managerUserId: string): Promise<string> {
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:68]:
```suggestion
  async getWithdrawalUrl(managedUserId: string): Promise<string> {
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 4]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     const token = await this.getIframeAuthorizationToken(
       this.clientIds.onOffRamp,
-      DEFAULT_APP_SCOPE
+      DEFAULT_APP_SCOPE,
+      managerUserId
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:72]:
```suggestion
      managedUserId
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 5]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     )
 
     return `${this.rampUrl}/?paymentType=${PAYMENT_TYPE.withdrawal}&bearer=${token}`
   }
 
-  async getDepositUrl(): Promise<string> {
+  async getDepositUrl(managerUserId: string): Promise<string> {
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:78]:
```suggestion
  async getDepositUrl(managedUserId: string): Promise<string> {
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 6]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     const token = await this.getIframeAuthorizationToken(
       this.clientIds.onOffRamp,
-      DEFAULT_APP_SCOPE
+      DEFAULT_APP_SCOPE,
+      managerUserId
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:82]:
```suggestion
      managedUserId
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 7]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     return `${this.rampUrl}/?paymentType=${PAYMENT_TYPE.deposit}&bearer=${token}`
   }
 
-  async getOnboardingUrl(): Promise<string> {
+  async getOnboardingUrl(managerUserId: string): Promise<string> {
     const token = await this.getIframeAuthorizationToken(
       this.clientIds.onboarding,
-      ONBOARDING_APP_SCOPE
+      ONBOARDING_APP_SCOPE,
+      managerUserId
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:92]:
```suggestion
      managedUserId
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 8]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     )
 
     return `${this.onboardingUrl}/?bearer=${token}`
   }
 
-  async getExchangeUrl(): Promise<string> {
+  async getExchangeUrl(managerUserId: string): Promise<string> {
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:98]:
```suggestion
  async getExchangeUrl(managedUserId: string): Promise<string> {
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 9]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     const token = await this.getIframeAuthorizationToken(
       this.clientIds.exchange,
-      DEFAULT_APP_SCOPE
+      DEFAULT_APP_SCOPE,
+      managerUserId
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:102]:
```suggestion
      managedUserId
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 10]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
     )
 
     return `${this.exchangeUrl}/?bearer=${token}`
   }
 
-  async getIframeUrl(type: IFRAME_TYPE): Promise<string> {
+  async getIframeUrl(
+    type: IFRAME_TYPE,
+    managerUserId: string
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:110]:
```suggestion
    managedUserId: string
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 11]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
+  ): Promise<string> {
     if (!this.iframeMappings[type]) {
       throw new BadRequest('Invalid iframe type')
     }
 
-    return await this.iframeMappings[type]()
+    return await this.iframeMappings[type](managerUserId)
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:116]:
```suggestion
    return await this.iframeMappings[type](managedUserId)
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 12]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
   }
 
   async getIframeAuthorizationToken(
     clientId: string,
-    scope: string[]
+    scope: string[],
+    managerUserId: string
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:122]:
```suggestion
    managedUserId: string
```

File: packages/wallet/backend/src/gatehub/client.ts [hunk 13]
```diff
@@ -58,61 +65,70 @@ export class GateHubClient {
...
   ): Promise<string> {
     const url = `${this.apiUrl}/auth/v1/tokens?${clientId}`
     const body: ITokenRequest = { scope }
 
     const response: ITokenResponse = await this.request<ITokenResponse>(
       'POST',
       url,
-      JSON.stringify(body)
+      JSON.stringify(body),
+      managerUserId
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/client.ts:131]:
```suggestion
      managedUserId
```

File: packages/wallet/backend/src/gatehub/service.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+export class GateHubService {
+  constructor(private gateHubClient: GateHubClient) {}
+
+  async getIframeUrl(iframeType: IFRAME_TYPE, userId: string): Promise<string> {
+    const user = await User.query().findById(userId)
+    if (!user || !user.gateHubUserId) {
+      throw new NotFound()
+    }
+
+    const url = this.gateHubClient.getIframeUrl(iframeType, user.gateHubUserId)
```
Comment (raducristianpopa) [inline packages/wallet/backend/src/gatehub/service.ts:15]:
```suggestion
    const url = await this.gateHubClient.getIframeUrl(iframeType, user.gateHubUserId)
```

File: packages/wallet/backend/tests/incomingPayment/service.test.ts [hunk 15]
```diff
@@ -30,7 +30,7 @@ describe('Incoming Payment Service', () => {
       assetCode: mockedListAssets[0].code,
       assetId: mockedListAssets[0].id,
       assetScale: mockedListAssets[0].scale,
-      virtualAccountId: 'mocked'
+      gateHubUserId: 'mocked'
```
Comment (raducristianpopa) [inline packages/wallet/backend/tests/incomingPayment/service.test.ts:33]: Mostly for all the tests, this should be `gateHubWalletId`
```suggestion
      gateHubWalletId: 'mocked'
```

File: packages/wallet/backend/src/account/service.ts [hunk 16]
```diff
@@ -70,51 +61,25 @@ export class AccountService implements IAccountService {
...
-      throw new Error(
-        `Unable to issue virtal account to ewallet: ${result.status?.message}`
-      )
-    }
-    // save virtual bank account number to database
-    const virtualAccount = result.data
+    const result = await this.gateHubClient.createWallet(
+      user.gateHubUserId,
+      args.name
+    )
```
[the remark this answers is not in the document]
Comment (dragosp1011) [reply packages/wallet/backend/src/account/service.ts:71]: should we use one wallet for all the assets?

[the remark this answers is not in the document]
Comment (dragosp1011) [reply packages/wallet/backend/src/account/service.ts:71]: I thought that a wallet for each currency may be easier to manage

[the remark this answers is not in the document]
Comment (raducristianpopa) [reply packages/wallet/backend/src/account/service.ts:71]: > should we use one wallet for all the assets?

CC @beniaminmunteanu @sanducb

[the remark this answers is not in the document]
Comment (sanducb) [reply packages/wallet/backend/src/account/service.ts:71]: IMO having one wallet for each currency indeed seems easier to manage (as suggested by @dragosp1011). Also, looking into the GateHub [docs](https://docs.gatehub.net/api-documentation/c3OPAp5dM191CDAdwyYS/api-reference/api-reference/wallets):
```
Wallets are accounts that can hold one or more assets.
```
I would say that since for GateHub wallets are considered accounts, then we could use the same account abstraction on our side and respect our own rules i.e. having one currency per account.

File: packages/wallet/backend/migrations/20230317135105_create_users.js [hunk 17]
```diff
@@ -17,6 +17,8 @@ exports.up = function (knex) {
     table.string('rapydContactId')
     table.string('kycId')
```
Comment (sanducb) [inline packages/wallet/backend/migrations/20230317135105_create_users.js:18]: I assume we will delete the Rapyd related columns in a different PR?

File: packages/wallet/backend/src/account/service.ts [hunk 18]
```diff
@@ -70,51 +61,25 @@ export class AccountService implements IAccountService {
...
-    const accountsBalance = await this.rapydClient.getAccountsBalance(
-      user.rapydWalletId
-    )
-
-    account.balance = transformBalance(
-      accountsBalance.data?.find((acc) => acc.currency === account.assetCode)
-        ?.balance ?? 0,
-      asset.scale
-    )
+    account.balance = transformBalance(0, account.assetScale)
```
Comment (sanducb) [inline packages/wallet/backend/src/account/service.ts:82]: Looks like this will also be handled in a different PR as well? Or should the balance be 0 at the account creation?

**Review by sanducb — COMMENTED** [review]
Looks good to me overall, I only left some questions

File: packages/wallet/backend/migrations/20230317135105_create_users.js (hunk 17 above)
Comment (dragosp1011) [reply packages/wallet/backend/migrations/20230317135105_create_users.js:18]: yes, deleted most of the rapyd related columns [here](https://github.com/interledger/testnet/pull/1612)
kycId one it is kept as will be replaced with GateHub info if kyc was completed

File: packages/wallet/backend/src/account/service.ts (hunk 18 above)
Comment (dragosp1011) [reply packages/wallet/backend/src/account/service.ts:82]: at account creation the balance is always 0
